Fix concurrency issues with cache

nateprewitt opened the pull request:
## JMESPath Cache Fixes

This PR attempts to address two issues with the existing cache implementation. The first is an issue with how we choose static sizing for cache eviction which means we may not clear the cache back to a reasonable size without multiple calls.
The second is a race condition in the cache that appeared in Python 3.10+, causing cache eviction to fail with `Sample larger than population or is negative`.

### Cache Free Size Issues
Currently the cache performs `self._MAX_SIZE/2` (64) evictions on random keys in the cache when the size exceeds `self._MAX_SIZE` (128). The issue with this is in highly parallelized run times, the cache can significantly exceed the size limit. The next invocation on `parse()` only clears a static amount, which means evicting from the cache may take several cycles.

### Cache Eviction Race Condition

When `parse()` is called in a highly parallelized environment, it's possible for two invocations to hit the cache clear at the same time. If some clears finish before another begins, we can invoke `random.sample` with a `k` (in this case 64) larger than our cache. That results in a [ValueError](https://github.com/python/cpython/blob/726e8e8defc487c55ade9f6015aaf2bae6cec3f3/Lib/random.py#L435-L436) because we're requesting more samples than we have keys.

## Updated Fix (Post-discussion)
This PR now changes the MAX_SIZE to 512 allowing for a larger cache to avoid needing eviction in the common case. It also switches the cache to a simpler FIFO implementation that will only affect one entry at a time. This should have minimal impact on performance from initial testing and ensure we don't hit a race condition without the lock. In the event that an entry cannot be evicted due another thread removing the entry, it will be treated as a no-op.

nateprewitt committed:
Demonstrate failure in existing test suite

jamesls commented:
Thanks for the PR.  Given how long this code's been around I think it's worth validating the original assumptions still hold.  I'd like to check:

1. What is the distribution of parsing times for "typical" (whatever that might mean) expressions without any caching
2. What, if any, perf improvements are there with the existing caching scheme in a single threaded scenario.
3. What does the perf look like with the introduction of a lock in the single threaded case.
4. What does the perf look like with lru_cache used instead

I suppose we'll need to choose some type of realistic distribution that approximates hot keys with some long tail of cold keys. 
I'm really curious to see how helpful a cache is in modern times.  I'm fairly confident we can do this without any locking.  I also wouldn't be surprised if the caching no longer has any tangible benefits.

jamesls commented:
Spent some time fixing up some of the old code and benchmarks in `./perf`.  There's some interesting takeaways here that somewhat confirm what I suspected.  On the one hand, caching does speed things up:

- Caching provides 20x to 4500x speedup on parse times when cache hits occur
- However, parse times are already in the microsecond range (`1.4-320μs`) without caching, `0.06μs` for a cache hit, using the default `tests/compliance/benchmarks.json` file.
- The lock introduced in this PR has negligible performance impact compared to the no-lock approach for the *single-threaded* case (no benchmarks for multi-threaded case).

But again we're talking microseconds...in Python.

I went through the git logs and found in commit 0835c1eba492d2e36c1ea47f6dfee2e6a92c1ef5, we had these timings:

**With original third party parsing library**

```
lex_time: 1.11579895, parse_time: 1.56807899ms name: deep_nesting_10
lex_time: 1.52611732, parse_time: 2.52008438ms name: deep_nesting_50
```

**With hand written parser**

```
lex_time: 0.07581711, parse_time: 0.14400482ms name: deep_nesting_10
lex_time: 0.37312508, parse_time: 0.69999695ms name: deep_nesting_50
```

And after fixing up the perf cases to work with the latest develop branch, on my m2 macbook pro, I'm seeing:

**2025 m2 macbook**

```
lex_time:    6.81865us, parse_time:   15.28971us name: deep_nesting_10
lex_time:   32.86454us, parse_time:   77.18332us name: deep_nesting_50
```

To be clear that's almost 9 times faster with today's hardware (700us vs. 77.18us), and 33x faster compared to the original original parser (2520us vs. 77.18us).  I'm not sure we really care as much about caching as we did 11+ years ago.

So what I'd propose:

* throw away the existing cache purging algorithm
* remove the lock
* bump the cache size up to 512
* purge only a single entry if cache size exceeded
* use a basic FIFO eviction, don't bother with LRU

Thoughts?

nateprewitt committed:
Add fifo queue impl

jamesls committed:
Implement simpler fifo cache

This simplifies the caching logic while addressing the thread safety
issue when purging the cache.

Caching parsed expressions is essentially a historical artifact at this
point.  The parse times are on the order of microseconds now, so while
caching does improve performance, no one's going to notice microseconds
in Python.

As a compromise, the caching is left in place, with a simpler
implementation:

* A single element is purged if we exceed max cache size
* Bump the max cache size from 128 to 512
* Use basic FIFO eviction, relying on py37+ dict insertion order

Ran the failing test with 30 threads and `1_000_000` requests and
confirmed we don't see any errors and that the `self._MAX_SIZE` is
maintained.

nateprewitt commented:
That seems reasonable if we want to maintain the cache ourselves. I think there's still technically a race condition but we can get around that by making a missing key as a no-op now that we're only clearing one entry at a time. I can start on that if you want or if you've got a specific design I'm happy to review.

jamesls committed:
Add changelog entry for bugfix

jamesls reviewed (approved):
Cleaned up my comments and commit messages, and added a changelog entry, looks good to me.  Thanks!
